target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,219 @@
+#include <linux/of.h>
+#include <linux/platform_device.h>
+
+#include "rb_hardconfig.h"
+#include "routerboot.h"
+
+#define TLV_TAG_MASK 	GENMASK(15, 0)
+#define TLV_LEN_MASK 	GENMASK(31, 16)
+
+static const char *rb_tlv_cell_name(u16 tag)
+{
+	switch (tag) {
+	case RB_ID_FLASH_INFO:
+		return "flash-info";
+	case RB_ID_MAC_ADDRESS_PACK:
+		return "base-mac-address";
+	case RB_ID_BOARD_PRODUCT_CODE:
+		return "board-product-code";
+	case RB_ID_BIOS_VERSION:
+		return "booter-version";
+	case RB_ID_SERIAL_NUMBER:
+		return "board-serial";
+	case RB_ID_MEMORY_SIZE:
+		return "mem-size";
+	case RB_ID_MAC_ADDRESS_COUNT:
+		return "mac-count";
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
+	case RB_ID_BOARD_IDENTIFIER:
+		return "board-identifier";
+	case RB_ID_PRODUCT_NAME:
+		return "product-name";
+	case RB_ID_DEFCONF:
+		return "defconf";
+	case RB_ID_BOARD_REVISION:
+		return "board-revision";
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_tlv_mac_read_cb(void *priv, const char *id, int index,
+			      unsigned int offset, void *buf,
+			      size_t bytes)
+{
+	eth_addr_add(buf, index);
+
+	return 0;
+}
+
+static nvmem_cell_post_process_t rb_tlv_read_cb(u16 tag, u8 *buf)
+{
+	switch (tag) {
+	case RB_ID_MAC_ADDRESS_PACK:
+		return &rb_tlv_mac_read_cb;
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_add_cells(struct device *dev, struct nvmem_device *nvmem,
+			const size_t magic_hdr_len, const size_t data_len,
+			u8 *data)
+{
+	u32 node, offset = 0, node_len = sizeof(node);
+	struct nvmem_cell_info cell = {};
+	struct device_node *layout;
+	u16 tlv_tag, tlv_len;
+	int ret;
+
+	layout = of_nvmem_layout_get_container(nvmem);
+	if (!layout)
+		return -ENOENT;
+
+	/*
+	 * Routerboot tag nodes are u32 values:
+	 * - The low nibble is the tag identification number,
+	 * - The high nibble is the tag payload length (node excluded) in bytes.
+	 * Tag nodes are CPU-endian.
+	 * Tag nodes are 32bit-aligned.
+	 *
+	 * The payload immediately follows the tag node.
+	 * Payload offset will always be aligned. while length may not end on 32bit
+	 * boundary (the only known case is when parsing ERD data).
+	 * The payload is CPU-endian when applicable.
+	 * Tag nodes are not ordered (by ID) on flash.
+	 */
+	while (offset < data_len) {
+		node = *((const u32 *) (data + offset));
+		/* Tag list ends with null node */
+		if (!node)
+			break;
+
+		tlv_tag = FIELD_GET(TLV_TAG_MASK, node);
+		tlv_len = FIELD_GET(TLV_LEN_MASK, node);
+
+		cell.name = rb_tlv_cell_name(tlv_tag);
+		if (!cell.name)
+			goto skip;
+
+		cell.offset = magic_hdr_len + offset + node_len;
+		/*
+		 * MikroTik stores MAC-s with length of 8 bytes,
+		 * but kernel expects it to be ETH_ALEN (6 bytes),
+		 * so we need to make sure that is the case.
+		 */
+		if (tlv_tag == RB_ID_MAC_ADDRESS_PACK)
+			cell.bytes = ETH_ALEN;
+		else
+			cell.bytes = tlv_len;
+		cell.np = of_get_child_by_name(layout, cell.name);
```

Comment:
Wants `of_node_put`: https://docs.kernel.org/devicetree/kernel-api.html#c.of_get_child_by_name
```suggestion
		cell.np = of_get_child_by_name(layout, cell.name);
		of_node_put(cell.np);
```

Reply:
Hm, I am not sure that `of_node_put` should be used here as that node can/will be used later.
I know that other layout drivers do not use `of_node_put` on the cell node.